File: Linked Lists/reorder list.py
class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self) -> None:
        self.head = None

    def append(self, val):
        new_node = Node(val)

        if not self.head:
            self.head = new_node
            return
        
        curr = self.head
        while curr.next:
            curr = curr.next
        curr.next = new_node
        
    def print_list(self):
        curr = self.head

        while curr:
            print(curr.data,end=" -> ")
            curr = curr.next
        print('None')


    def reorder_list(self, head):
        slow = fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        second = slow.next
        print(slow.data)

        prev = slow.next = None
        
        while second:
            next_node = second.next
            second.next = prev
            prev = second
            second = next_node

        first = head
        second = prev

        while second:
            temp1 = first.next
            temp2 = second.next

            first.next = second
            second.next = temp1

            first = temp1
            second = temp2

lst = LinkedList()
arr = [1,2,3,4,5,6]

for i in arr:
    lst.append(i)

lst.print_list()
lst.reorder_list(lst.head)
lst.print_list()

File: CPP/Linked Lists/reorder list.cpp
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int data;
    Node *next;

public:
    Node(int val1)
    {
        data = val1;
        next = nullptr;
    }

public:
    Node(int val1, Node *next1)
    {
        data = val1;
        next = next1;
    }
};

Node *convertArr2LL(vector<int> arr)
{
    Node *head = new Node(arr[0]);
    Node *curr = head;

    for (int i = 1; i < arr.size(); i++)
    {
        Node *temp = new Node(arr[i]);

        curr->next = temp;
        curr = curr->next;
    }

    return head;
}

void print(Node *head)
{
    Node *curr = head;

    while (curr)
    {
        cout << curr->data << " -> ";
        curr = curr->next;
    }
    cout << "null" << endl;
}

Node *reverse(Node *head)
{
    Node *curr = head, *prev = nullptr, *nextNode = nullptr;

    while (curr)
    {
        nextNode = curr->next;
        curr->next = prev;
        prev = curr;
        curr = nextNode;
    }
    return prev;
}

Node *reorderList(Node *head)
{
    if (!head || !head->next)
        return head;

    Node *dummy = new Node(0, head);
    Node *first = head, *slow = head, *fast = head;

    while (fast->next && fast->next->next)
    {
        slow = slow->next;
        fast = fast->next->next;
    }

    Node *second = slow->next;
    slow->next = nullptr;

    second = reverse(second);

    while (second)
    {
        Node *temp1 = first->next;
        Node *temp2 = second->next;

        first->next = second;
        second->next = temp1;

        first = temp1;
        second = temp2;
    }

    return dummy->next;
}

int main()
{
    vector<int> arr = {1, 2, 3, 4, 5, 6};
    Node *head = convertArr2LL(arr);
    print(head);

    head = reorderList(head);
    print(head);

    return 0;
}